### DISCUSSION TITLE
How UEFI selects the image consistent with its architecture from the option rom, loads and executes it?

### DISCUSSION THREAD

**Discussion by xuchao1234** [tianocore/edk2#4048] [Q&A] [answered]
I didn't find any relevant processing in the "ProcessOpRomImage" function. Please tell me the location of the relevant code, Thank very much!

**Comment by Kostr** [comment, marked as answer]
Hi!

I'm not an expert, but I've tried to investigate this subject myself.

According to the UEFI specification target `Machine Type` is encoded in the header of a ROM image:
![2023-02-15_21-36-22](https://user-images.githubusercontent.com/2064127/219128023-c9ee1fa9-befa-46ad-b089-17915f1c5e4b.png)

You can find this field in the relevant `EFI_PCI_EXPANSION_ROM_HEADER` structure https://github.com/tianocore/edk2/blob/master/MdePkg/Include/IndustryStandard/Pci22.h

```cpp
///
/// EFI PCI Expansion ROM Header
/// Section 13.4.2, Unified Extensible Firmware Interface Specification, Version 2.1
///
typedef struct {
  UINT16    Signature;  ///< 0xaa55
  UINT16    InitializationSize;
  UINT32    EfiSignature; ///< 0x0EF1
  UINT16    EfiSubsystem;
  UINT16    EfiMachineType;                      // <-------------
  UINT16    CompressionType;
  UINT8     Reserved[8];
  UINT16    EfiImageHeaderOffset;
  UINT16    PcirOffset;
} EFI_PCI_EXPANSION_ROM_HEADER;
```

But it is not processed neither in the PCI ROM parsing code, nor in the `loadpcirom` UEFI shell command code:

https://github.com/tianocore/edk2/blob/7c0ad2c33810ead45b7919f8f8d0e282dae52e71/MdeModulePkg/Bus/Pci/PciBusDxe/PciOptionRomSupport.c#L76

https://github.com/tianocore/edk2/blob/7c0ad2c33810ead45b7919f8f8d0e282dae52e71/ShellPkg/Library/UefiShellDebug1CommandsLib/LoadPciRom.c#L268

So it looks like EDKII just loads every image without checking its machine type. And it is responsibility of the `LoadImage()` function to fail if the architecture is wrong.

Here is the relevant info from the UEFI specification that supports this conclusion:
![2023-02-15_22-09-41](https://user-images.githubusercontent.com/2064127/219129328-92b18c79-0b78-4f24-99f7-5a5ef7ff5aa9.png)
